pkg/apis/networking/validation/validation.go

```diff
@@ -174,15 +179,35 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
 	return allErrs
 }
 
-// ValidateIngress tests if required fields in the Ingress are set.
+// ValidateIngressName validates that the given name can be used as an Ingress
+// name.
+var ValidateIngressName = apimachineryvalidation.NameIsDNSSubdomain
+
+// ValidateIngress validates Ingresses on create and update.
 func ValidateIngress(ingress *networking.Ingress) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMeta(&ingress.ObjectMeta, true, ValidateIngressName, field.NewPath("metadata"))
+	annotationVal, annotationIsSet := ingress.Annotations[annotationIngressClass]
+	if annotationIsSet && ingress.Spec.IngressClassName != nil {
```

Comment:
Good catch, you're right. I'd made this change after your recommendation but failed to update the godoc to match that. Before I do that, one bit of clarification, is there a straightforward path to remove an annotation and set a field in one step with something like `kubectl`? It seems like you could use a relatively complex `patch`, but I can't think of a way to accomplish this with something more common like `kubectl apply`. It almost seems like it would be better to allow for updates to allow the class field to be set if the annotation is already set, just for ease of upgrading.

Reply:
if it's well-specified what should happen when both are defined, I don't particularly see a reason to block it. I'd expect it to be pretty normal to see Ingress manifests that try to set both the field and the annotation for a while, to work against clusters newer and older than 1.18

Reply:
I think the most compelling reason to set both the field and annotation is to simplify the transition from one to the other. If the annotation is always trusted over the class, the value here is just enabling that transition from annotation to field to be a little smoother. With that background, I'm not sure why anyone would want to create a new Ingress with both fields set. I'm leaning towards moving back to what the KEP specifies - only one of the field or annotation can be set on creates, both are valid for updates. Does that sound reasonable?

Reply:
That requires manifest authors to target manifests at a specific cluster version (and maintain different manifests for < 1.18 and >= 1.18, which is pretty tedious. I could see restricting that for v1.

Reply:
The class field is only being added in 1.18 though, so this would only be possible >= 1.18 right? With that said, I don't mind leaving it open on both create and update now and revisiting this for v1/1.19.

Reply:
For now I've gone back to what we had in the KEP - validation prevents both fields being set on new Ingresses. I'm with switching this to leaving it open fully though if that's needed for better interoperability between versions, just let me know.

Reply:
ok